File: mavsim_matlab/chap12/path_planner.m
% path planner for mavsim_matlab
%
% mavsim_matlab 
%     - Beard & McLain, PUP, 2012
%     - Last updated:  
%         4/2/2019 - RWB
classdef path_planner < handle
   %--------------------------------
    properties
        waypoints
        rrt_straight_line
        rrt_dubins
    end
    %--------------------------------
    methods
        %------constructor-----------
        function self = path_planner
            % waypoints definition
            addpath('../message_types'); 
            self.waypoints = msg_waypoints(); 
            addpath('../chap12')
            self.rrt_straight_line = rrt_straight_line();
            self.rrt_dubins = rrt_dubins();
        end
        %------methods-----------
        function waypoints = update(self, map, state, radius)
            % this flag is set for one time step to signal a redraw in the
            % viewer
            %planner_flag = 1;  % return simple waypoint path
            %planner_flag = 2;  % return dubins waypoint path
            %planner_flag = 3;  % plan path through city using straight-line RRT
            planner_flag = 4;  % plan path through city using dubins RRT
            disp('planning...');
            if planner_flag == 1
                Va = 25;
                self.waypoints.type = 'fillet';
                self.waypoints.add([0; 0; -100], Va, inf, inf, 0, 0);
                self.waypoints.add([1000; 0; -100], Va, inf, inf, 0, 0);
                self.waypoints.add([0; 1000; -100], Va, inf, inf, 0, 0);
                self.waypoints.add([1000; 1000; -100], Va, inf, inf, 0, 0);
            elseif planner_flag == 2
                Va = 25;
                self.waypoints.type = 'dubins';
                self.waypoints.add([0; 0; -100], Va, 0, inf, 0, 0);
                self.waypoints.add([1000; 0; -100], Va, 45*pi/180, inf, 0, 0);
                self.waypoints.add([0; 1000; -100], Va, 45*pi/180, inf, 0, 0);
                self.waypoints.add([1000; 1000; -100], Va, -135*pi/180, inf, 0, 0);
            elseif planner_flag == 3
                Va = 25;
                % start configuration is current configuration
                ps = [state.pn; state.pe; -state.h];
                % desired end configuration
                if norm(ps -[map.city_width; map.city_width; -state.h]) < map.city_width/2
                    pe = [0; 0; -state.h];
                else
                    pe = [map.city_width; map.city_width; -state.h];
                end
                self.waypoints.type = 'fillet';
                self.waypoints = self.rrt_straight_line.update(ps, pe, Va, map, radius);
            elseif planner_flag == 4
                Va = 25;
                % start pose is current pose
                start_pose = [state.pn; state.pe; -state.h; state.chi];
                % desired end pose
                if norm(start_pose(1:3) -[map.city_width; map.city_width; -state.h]) < map.city_width/2
                    end_pose = [0; 0; -state.h; 0];
                else
                    end_pose = [map.city_width; map.city_width; -state.h; 0];
                end
                self.waypoints.type = 'dubins';
                self.waypoints = self.rrt_dubins.update(start_pose, end_pose, Va, map, radius);
            else
                disp('Error in Path Planner: Undefined planner type.')
            end
            % return the planned waypoints
            waypoints = self.waypoints;
            disp('... done planning');
        end
    end
end

File: mavsim_python_legacy/chap12/path_planner.py
# path planner for mavsim_python
#
# mavsim_python
#     - Beard & McLain, PUP, 2012
#     - Last updated:
#         4/3/2019 - BGM
import numpy as np
import sys
sys.path.append('..')
from message_types.msg_waypoints import MsgWaypoints
from chap12.rrt_straight_line import RRTStraightLine
from chap12.rrt_dubins import RRTDubins


class PathPlanner:
    def __init__(self):
        # waypoints definition
        self.waypoints = MsgWaypoints()
        self.rrt_straight_line = RRTStraightLine()
        self.rrt_dubins = RRTDubins()

    def update(self, world_map, state, radius):
        print('planning...')
        # this flag is set for one time step to signal a redraw in the viewer
        # planner_flag = 'simple_straight'  # return simple waypoint path
        # planner_flag = 'simple_dubins'  # return simple dubins waypoint path
        planner_flag = 'rrt_straight'  # plan path through city using straight-line RRT
        #planner_flag = 'rrt_dubins'  # plan path through city using dubins RRT
        if planner_flag == 'simple_straight':
            Va = 25
            self.waypoints.type = 'fillet'
            self.waypoints.add(np.array([[0, 0, -100]]).T, Va, np.inf, np.inf, 0, 0)
            self.waypoints.add(np.array([[1000, 0, -100]]).T, Va, np.inf, np.inf, 0, 0)
            self.waypoints.add(np.array([[0, 1000, -100]]).T, Va, np.inf, np.inf, 0, 0)
            self.waypoints.add(np.array([[1000, 1000, -100]]).T, Va, np.inf, np.inf, 0, 0)

        elif planner_flag == 'simple_dubins':
            Va = 25
            self.waypoints.type = 'dubins'
            self.waypoints.add(np.array([[0, 0, -100]]).T, Va, np.radians(0), np.inf, 0, 0)
            self.waypoints.add(np.array([[1000, 0, -100]]).T, Va, np.radians(45), np.inf, 0, 0)
            self.waypoints.add(np.array([[0, 1000, -100]]).T, Va, np.radians(45), np.inf, 0, 0)
            self.waypoints.add(np.array([[1000, 1000, -100]]).T, Va, np.radians(-135), np.inf, 0, 0)

        elif planner_flag == 'rrt_straight':
            desired_airspeed = 25
            desired_altitude = 100
            # start pose is current pose
            start_pose = np.array([[state.north], [state.east], [-desired_altitude]])
            # desired end pose
            if np.linalg.norm(start_pose[0:2]) < world_map.city_width / 2:
                end_pose = np.array([[world_map.city_width], [world_map.city_width],
                                     [-desired_altitude]])
            else:  # or to the bottom-left corner of world_map
                end_pose = np.array([[0], [0], [-desired_altitude]])
            self.waypoints = self.rrt_straight_line.update(start_pose, end_pose,
                                                           desired_airspeed, world_map, radius)

        elif planner_flag == 'rrt_dubins':
            desired_airspeed = 25
            desired_altitude = 100
            # start pose is current pose
            start_pose = np.array([[state.north], [state.east],
                                   [-desired_altitude], [state.chi]])
            # desired end pose
            # either plan to the top-right corner of world_map
            if np.linalg.norm(start_pose[0:2]) < world_map.city_width / 2:
                end_pose = np.array([[world_map.city_width], [world_map.city_width],
                                     [-desired_altitude], [state.chi]])
            else:  # or to the bottom-left corner of world_map
                end_pose = np.array([[0], [0], [-desired_altitude], [state.chi]])
            self.waypoints = self.rrt_dubins.update(start_pose, end_pose,
                                                    desired_airspeed, world_map, radius)
        else:
            print("Error in Path Planner: Undefined planner type.")
        self.waypoints.plot_updated = False
        print('...done planning.')
        return self.waypoints
